osquery/tables/system/windows/registry.cpp

```diff
@@ -266,33 +269,145 @@ void queryKey(const std::string& keyPath, QueryData& results) {
+  std::set<std::string> newKeys{};
+  for (auto& key : rKeys) {
+    newKeys.insert(key + kRegSep + subkey);
+  }
+  rKeys = newKeys;
+}
+
+Status populateAllKeysRecursive(std::set<std::string>& rKeys,
+                                int currDepth,
+                                int maxDepth) {
+  std::set<std::string> subkeys{};
+
+  if (currDepth > maxDepth) {
+    return Status(
+        1, "Max recursive depth (" + std::to_string(maxDepth) + ") reached");
+  }
+
+  for (const auto& key : rKeys) {
+    QueryData regResults;
+    queryKey(key, regResults);
+    for (const auto& r : regResults) {
+      if (r.at("type") == "subkey") {
+        subkeys.insert(r.at("path"));
+      }
+    }
+  }
+
+  if (!(subkeys.size() == 0)) {
+    auto status = populateAllKeysRecursive(subkeys, ++currDepth);
+    if (!status.ok()) {
+      return status;
+    }
+    rKeys.insert(subkeys.begin(), subkeys.end());
+  }
+
+  return Status(0, "OK");
+}
+
+Status resolveRegistryGlobs(const std::string& pattern,
+                            std::set<std::string>& results) {
+  auto pathElems = osquery::split(pattern, kRegSep);
+
+  // Pattern is '%%', grab everything
+  // Note that if '%%' is present but not at the end of the pattern
+  // , then it is treated like a single glob
+  if (boost::ends_with(pathElems[0], kRegRecursiveGlob) && pathElems.size() == 1) {
+    populateDefaultKeys(results);
+    return populateAllKeysRecursive(results);
+  }
+  
+  // Special handling to insert default keys when glob present in first elem
+  if (pathElems[0].find(kRegSingleGlob) != std::string::npos) {
+    populateDefaultKeys(results);
+    pathElems.erase(pathElems.begin());
+  }
+  else {
+    results.insert(pathElems[0]);
+    pathElems.erase(pathElems.begin());
+  }
+
+  for (const auto& elem : pathElems) {
+     // We only care about  a recursive glob if it comes at the end of the pattern
+     // i.e. 'HKEY_LOCAL_MACHINE\SOFTWARE\%%'
+    if (boost::ends_with(elem, kRegRecursiveGlob) && &elem == &pathElems.back()) {
+      return populateAllKeysRecursive(results);
+    } else if (elem.find(kRegSingleGlob) != std::string::npos) {
+      replaceKeysWithSubkeys(results);
+    } else {
+        appendSubkeyToKeys(elem, results);
     }
   }
+  return Status(0, "OK");
+}
 
+void maybeWarnLocalUsers(const std::set<std::string>& rKeys) {
+  std::string hive, _;
   for (const auto& key : rKeys) {
-    std::string hive;
-    std::string keyPath;
-    explodeRegistryPath(key, hive, keyPath);
-    if (shouldWarnLocalUsers && (hive == "HKEY_CURRENT_USER" ||
-                                 hive == "HKEY_CURRENT_USER_LOCAL_SETTINGS")) {
+    explodeRegistryPath(key, hive, _);
+    if (hive == "HKEY_CURRENT_USER" ||
+        hive == "HKEY_CURRENT_USER_LOCAL_SETTINGS") {
       LOG(WARNING) << "CURRENT_USER hives are not queryable by osqueryd; "
                       "query HKEY_USERS with the desired users SID instead";
-      shouldWarnLocalUsers = false;
+      return;
```

Comment:
Instead of having multiple returns from this method, maybe `break`, defer to your style preference.